Give the interrupt fault page a whole host page and check for suspends on conditional back-edges.

diff --git a/FEXCore/Source/Interface/Core/Core.cpp b/FEXCore/Source/Interface/Core/Core.cpp
--- a/FEXCore/Source/Interface/Core/Core.cpp
+++ b/FEXCore/Source/Interface/Core/Core.cpp
@@ -361,6 +361,10 @@ bool ContextImpl::InitCore() {
 #if defined(_WIN32) && !defined(ARCHITECTURE_arm64ec)
   // WOW64 always needs the interrupt fault check to be enabled.
   Config.NeedsPendingInterruptFaultCheck = true;
+  if (FEXCore::Utils::GetHostPageSize() > FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE) {
+    LogMan::Msg::EFmt("Host page size {} is larger than the interrupt fault page; suspending threads running JIT code will not work",
+                      FEXCore::Utils::GetHostPageSize());
+  }
 #endif
 
   if (Config.GdbServer) {
@@ -444,7 +448,7 @@ FEXCore::Core::InternalThreadState* ContextImpl::CreateThread(const FEXCore::Cor
 }
 
 void ContextImpl::DestroyThread(FEXCore::Core::InternalThreadState* Thread) {
-  FEXCore::Allocator::VirtualProtect(&Thread->InterruptFaultPage, sizeof(Thread->InterruptFaultPage),
+  FEXCore::Allocator::VirtualProtect(FEXCore::Core::GetInterruptFaultPage(Thread), FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE,
                                      Allocator::ProtectOptions::Read | Allocator::ProtectOptions::Write);
   delete Thread;
 }
diff --git a/FEXCore/Source/Interface/Core/Dispatcher/Dispatcher.cpp b/FEXCore/Source/Interface/Core/Dispatcher/Dispatcher.cpp
--- a/FEXCore/Source/Interface/Core/Dispatcher/Dispatcher.cpp
+++ b/FEXCore/Source/Interface/Core/Dispatcher/Dispatcher.cpp
@@ -277,7 +277,7 @@ void Dispatcher::EmitDispatcher() {
 
     // Trigger segfault if any deferred signals are pending
     strb(ARMEmitter::XReg::zr, STATE,
-         offsetof(FEXCore::Core::InternalThreadState, InterruptFaultPage) - offsetof(FEXCore::Core::InternalThreadState, BaseFrameState));
+         FEXCore::Core::INTERRUPT_FAULT_PAGE_FRAME_OFFSET);
 #endif
   };
 
diff --git a/FEXCore/Source/Interface/Core/JIT/BranchOps.cpp b/FEXCore/Source/Interface/Core/JIT/BranchOps.cpp
--- a/FEXCore/Source/Interface/Core/JIT/BranchOps.cpp
+++ b/FEXCore/Source/Interface/Core/JIT/BranchOps.cpp
@@ -249,6 +249,10 @@ DEF_OP(CondJump) {
 
   auto TrueTargetLabel = JumpTarget(Op->TrueBlock);
 
+  if (TrueTargetLabel->Backward.Location) {
+    EmitSuspendInterruptCheck();
+  }
+
   if (Op->FromNZCV) {
     b_OrRestart(MapCC(Op->Cond), TrueTargetLabel);
   } else {
diff --git a/FEXCore/Source/Interface/Core/JIT/JIT.cpp b/FEXCore/Source/Interface/Core/JIT/JIT.cpp
--- a/FEXCore/Source/Interface/Core/JIT/JIT.cpp
+++ b/FEXCore/Source/Interface/Core/JIT/JIT.cpp
@@ -768,7 +768,7 @@ void Arm64JITCore::EmitSuspendInterruptCheck() {
     // Trigger a fault if there are any pending interrupts
     // Used only for suspend on WIN32 at the moment
     constexpr size_t InterruptPageOffset =
-      offsetof(FEXCore::Core::InternalThreadState, InterruptFaultPage) - offsetof(FEXCore::Core::InternalThreadState, BaseFrameState);
+      FEXCore::Core::INTERRUPT_FAULT_PAGE_FRAME_OFFSET;
     if constexpr (InterruptPageOffset <= 32760) {
       str(ARMEmitter::XReg::zr, STATE, InterruptPageOffset);
     } else {
diff --git a/FEXCore/include/FEXCore/Debug/InternalThreadState.h b/FEXCore/include/FEXCore/Debug/InternalThreadState.h
--- a/FEXCore/include/FEXCore/Debug/InternalThreadState.h
+++ b/FEXCore/include/FEXCore/Debug/InternalThreadState.h
@@ -88,6 +88,19 @@ struct UnalignedExclusiveStore {
 };
 
 struct alignas(FEXCore::Utils::FEX_PAGE_SIZE) InternalThreadState : public FEXCore::Allocator::FEXAllocOperators {
+  void* operator new(size_t size) {
+    return FEXCore::Allocator::aligned_alloc(FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE, size);
+  }
+  void* operator new(size_t size, std::align_val_t) {
+    return FEXCore::Allocator::aligned_alloc(FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE, size);
+  }
+  void operator delete(void* ptr) {
+    FEXCore::Allocator::aligned_free(ptr);
+  }
+  void operator delete(void* ptr, std::align_val_t) {
+    FEXCore::Allocator::aligned_free(ptr);
+  }
+
   FEXCore::Core::CpuStateFrame* const CurrentFrame = &BaseFrameState;
 
   FEXCore::Context::Context* const CTX;
@@ -122,12 +135,23 @@ struct alignas(FEXCore::Utils::FEX_PAGE_SIZE) InternalThreadState : public FEXCo
   FEXCore::Core::CpuStateFrame BaseFrameState {};
 
   // Can be reprotected as RO to trigger an interrupt at generated code block entrypoints
-  alignas(FEXCore::Utils::FEX_PAGE_SIZE) uint8_t InterruptFaultPage[FEXCore::Utils::FEX_PAGE_SIZE];
+  alignas(FEXCore::Utils::FEX_PAGE_SIZE) uint8_t InterruptFaultPage[2 * FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE - FEXCore::Utils::FEX_PAGE_SIZE];
 };
 static_assert(std::is_standard_layout_v<FEXCore::Core::InternalThreadState>);
+
+constexpr size_t INTERRUPT_FAULT_PAGE_OFFSET =
+  (offsetof(InternalThreadState, InterruptFaultPage) + FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE - 1) &
+  ~(FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE - 1);
+static_assert(INTERRUPT_FAULT_PAGE_OFFSET + FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE <=
+                offsetof(InternalThreadState, InterruptFaultPage) + sizeof(InternalThreadState::InterruptFaultPage),
+              "Interrupt fault page does not fit in its backing member");
+
+constexpr size_t INTERRUPT_FAULT_PAGE_FRAME_OFFSET = INTERRUPT_FAULT_PAGE_OFFSET - offsetof(InternalThreadState, BaseFrameState);
 // Maximum unsigned-offset store range for fault page.
-static_assert(
-  (offsetof(FEXCore::Core::InternalThreadState, InterruptFaultPage) - offsetof(FEXCore::Core::InternalThreadState, BaseFrameState)) <= 65520,
-  "Fault page is outside of immediate range from CPU state");
+static_assert(INTERRUPT_FAULT_PAGE_FRAME_OFFSET <= 65520, "Fault page is outside of immediate range from CPU state");
+
+inline uint8_t* GetInterruptFaultPage(InternalThreadState* Thread) {
+  return reinterpret_cast<uint8_t*>(Thread) + INTERRUPT_FAULT_PAGE_OFFSET;
+}
 
 } // namespace FEXCore::Core
diff --git a/FEXCore/include/FEXCore/Utils/SignalScopeGuards.h b/FEXCore/include/FEXCore/Utils/SignalScopeGuards.h
--- a/FEXCore/include/FEXCore/Utils/SignalScopeGuards.h
+++ b/FEXCore/include/FEXCore/Utils/SignalScopeGuards.h
@@ -138,12 +138,12 @@ public:
       // X86-64 must do an additional check around the store.
       if ((Result - 1) == 0) {
         // Must happen after the refcount store
-        auto InterruptFaultPage = reinterpret_cast<Core::NonAtomicRefCounter<uint64_t>*>(&Thread->InterruptFaultPage);
+        auto InterruptFaultPage = reinterpret_cast<Core::NonAtomicRefCounter<uint64_t>*>(Core::GetInterruptFaultPage(Thread));
         InterruptFaultPage->Store(0);
       }
 #else
       Thread->CurrentFrame->State.DeferredSignalRefCount.Decrement(1);
-      auto InterruptFaultPage = reinterpret_cast<Core::NonAtomicRefCounter<uint64_t>*>(&Thread->InterruptFaultPage);
+      auto InterruptFaultPage = reinterpret_cast<Core::NonAtomicRefCounter<uint64_t>*>(Core::GetInterruptFaultPage(Thread));
       InterruptFaultPage->Store(0);
 #endif
     }
diff --git a/FEXCore/include/FEXCore/Utils/TypeDefines.h b/FEXCore/include/FEXCore/Utils/TypeDefines.h
--- a/FEXCore/include/FEXCore/Utils/TypeDefines.h
+++ b/FEXCore/include/FEXCore/Utils/TypeDefines.h
@@ -12,5 +12,7 @@ constexpr size_t FEX_PAGE_MASK = ~(FEX_PAGE_SIZE - 1);
 
 constexpr size_t FEX_MAX_HOST_PAGE_SIZE = 65536;
 
+constexpr size_t FEX_INTERRUPT_FAULT_PAGE_SIZE = 16384;
+
 size_t GetHostPageSize();
 } // namespace FEXCore::Utils
diff --git a/Source/Tools/LinuxEmulation/LinuxSyscalls/SignalDelegator.cpp b/Source/Tools/LinuxEmulation/LinuxSyscalls/SignalDelegator.cpp
--- a/Source/Tools/LinuxEmulation/LinuxSyscalls/SignalDelegator.cpp
+++ b/Source/Tools/LinuxEmulation/LinuxSyscalls/SignalDelegator.cpp
@@ -419,7 +419,7 @@ bool SignalDelegator::HandleSIGILL(FEXCore::Core::InternalThreadState* Thread, i
       // If we have more deferred frames to process then mprotect back to PROT_NONE.
       // It will have been RW coming in to this sigreturn and now we need to remove permissions
       // to ensure FEX trampolines back to the SIGSEGV deferred handler.
-      mprotect(reinterpret_cast<void*>(&Thread->InterruptFaultPage), sizeof(Thread->InterruptFaultPage), PROT_NONE);
+      mprotect(FEXCore::Core::GetInterruptFaultPage(Thread), FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE, PROT_NONE);
     }
     return true;
   }
@@ -595,12 +595,12 @@ void SignalDelegator::HandleGuestSignal(FEX::HLE::ThreadStateObject* ThreadObjec
   auto SigInfo = *static_cast<siginfo_t*>(Info);
 
   auto MustDeferSignal = (Thread->CurrentFrame->State.DeferredSignalRefCount.Load() != 0);
-  if (Signal == SIGSEGV && SigInfo.si_code == SEGV_ACCERR && SigInfo.si_addr == reinterpret_cast<void*>(&Thread->InterruptFaultPage)) {
+  if (Signal == SIGSEGV && SigInfo.si_code == SEGV_ACCERR && SigInfo.si_addr == FEXCore::Core::GetInterruptFaultPage(Thread)) {
     if (!MustDeferSignal) {
       // We just reached the end of the outermost signal-deferring section and faulted to check for pending signals.
       // Pull a signal frame off the stack.
 
-      mprotect(reinterpret_cast<void*>(&Thread->InterruptFaultPage), sizeof(Thread->InterruptFaultPage), PROT_READ | PROT_WRITE);
+      mprotect(FEXCore::Core::GetInterruptFaultPage(Thread), FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE, PROT_READ | PROT_WRITE);
 
       if (ThreadObject->SignalInfo.DeferredSignalFrames.empty()) {
         // No signals to defer. Just set the fault page back to RW and continue execution.
@@ -659,7 +659,7 @@ void SignalDelegator::HandleGuestSignal(FEX::HLE::ThreadStateObject* ThreadObjec
     sigmask_val[0] = GetNewSigMask(Signal);
 
     // Now update the faulting page permissions so it will fault on write.
-    mprotect(reinterpret_cast<void*>(&Thread->InterruptFaultPage), sizeof(Thread->InterruptFaultPage), PROT_NONE);
+    mprotect(FEXCore::Core::GetInterruptFaultPage(Thread), FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE, PROT_NONE);
 
     // Postpone the remainder of signal handling logic until we process the SIGSEGV triggered by writing to InterruptFaultPage.
     return;
diff --git a/Source/Windows/WOW64/Module.cpp b/Source/Windows/WOW64/Module.cpp
--- a/Source/Windows/WOW64/Module.cpp
+++ b/Source/Windows/WOW64/Module.cpp
@@ -393,12 +393,12 @@ public:
 };
 
 bool HandleSuspendInterrupt(TLS TLS, CONTEXT* Context, uint64_t FaultAddress) {
-  if (FaultAddress != reinterpret_cast<uint64_t>(&TLS.ThreadState()->InterruptFaultPage)) {
+  if (FaultAddress != reinterpret_cast<uint64_t>(FEXCore::Core::GetInterruptFaultPage(TLS.ThreadState()))) {
     return false;
   }
 
   void* TmpAddress = reinterpret_cast<void*>(FaultAddress);
-  SIZE_T TmpSize = FEXCore::Utils::FEX_PAGE_SIZE;
+  SIZE_T TmpSize = FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE;
   ULONG TmpProt;
   NtProtectVirtualMemory(NtCurrentProcess(), &TmpAddress, &TmpSize, PAGE_READWRITE, &TmpProt);
 
@@ -836,8 +836,8 @@ NTSTATUS BTCpuSuspendLocalThread(HANDLE Thread, ULONG* Count) {
     LogMan::Msg::DFmt("Thread {:X} is in JIT, polling for interrupt", ThreadTID);
 
     ULONG TmpProt;
-    void* TmpAddress = &TLS.ThreadState()->InterruptFaultPage;
-    SIZE_T TmpSize = FEXCore::Utils::FEX_PAGE_SIZE;
+    void* TmpAddress = FEXCore::Core::GetInterruptFaultPage(TLS.ThreadState());
+    SIZE_T TmpSize = FEXCore::Utils::FEX_INTERRUPT_FAULT_PAGE_SIZE;
     NtProtectVirtualMemory(NtCurrentProcess(), &TmpAddress, &TmpSize, PAGE_READONLY, &TmpProt);
   }
 
